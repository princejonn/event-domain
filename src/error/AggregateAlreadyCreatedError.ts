import { DomainError } from "./DomainError";

export class AggregateAlreadyCreatedError extends DomainError {
  constructor() {
    super("Aggregate has already been created", { permanent: true });
  }
}
