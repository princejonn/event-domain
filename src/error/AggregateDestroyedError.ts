import { DomainError } from "./DomainError";

export class AggregateDestroyedError extends DomainError {
  constructor() {
    super("Aggregate is destroyed", { permanent: true });
  }
}
