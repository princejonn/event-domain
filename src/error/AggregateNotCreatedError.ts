import { DomainError } from "./DomainError";

export class AggregateNotCreatedError extends DomainError {
  constructor() {
    super("Aggregate has not been created", { permanent: true });
  }
}
