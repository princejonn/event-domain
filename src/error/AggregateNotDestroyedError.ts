import { DomainError } from "./DomainError";

export class AggregateNotDestroyedError extends DomainError {
  constructor() {
    super("Aggregate has not been destroyed", { permanent: true });
  }
}
