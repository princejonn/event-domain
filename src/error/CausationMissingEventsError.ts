import { LindormError } from "@lindorm-io/errors";

export class CausationMissingEventsError extends LindormError {
  constructor() {
    super("Causation produces no event array in aggregate");
  }
}
