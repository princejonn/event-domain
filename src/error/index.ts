export * from "./AggregateAlreadyCreatedError";
export * from "./AggregateDestroyedError";
export * from "./AggregateNotCreatedError";
export * from "./AggregateNotDestroyedError";
export * from "./CausationMissingEventsError";
export * from "./CommandSchemaValidationError";
export * from "./ConcurrencyError";
export * from "./DomainError";
export * from "./EventStreamRevisionError";
export * from "./HandlerNotRegisteredError";
export * from "./IllegalEntityChangeError";
export * from "./MessageTypeError";
export * from "./MongoDuplicateKeyError";
export * from "./MongoNotUpdatedError";
export * from "./SagaAlreadyCreatedError";
export * from "./SagaCausationError";
export * from "./SagaDestroyedError";
export * from "./SagaNotCreatedError";
export * from "./SagaNotInStoreError";
export * from "./SagaRevisionError";
export * from "./ViewAlreadyCreatedError";
export * from "./ViewCausationError";
export * from "./ViewDestroyedError";
export * from "./ViewNotCreatedError";
export * from "./ViewRevisionError";
