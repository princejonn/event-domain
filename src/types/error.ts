import { LindormErrorOptions } from "@lindorm-io/errors";

export interface DomainErrorOptions extends LindormErrorOptions {
  permanent: boolean;
}
