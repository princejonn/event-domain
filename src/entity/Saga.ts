import { Command, DomainEvent, Timeout } from "../message";
import {
  ISaga,
  SagaIdentifier,
  SagaOptions,
  AnyType,
  GenericRecord,
  SagaData,
} from "../types";
import { IllegalEntityChangeError, MessageTypeError, SagaDestroyedError } from "../error";
import { SAGA_DISPATCH_SCHEMA, SAGA_UPDATE_STATE_SCHEMA } from "../schema";
import { assertCamelCase, validateSchemaSync } from "../util";
import { cloneDeep, merge, set } from "lodash";
import { Logger } from "@lindorm-io/winston";

export class Saga<State> implements ISaga<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _causationList: Array<string>;
  private readonly _messagesToDispatch: Array<Command | DomainEvent>;
  private readonly _revision: number;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;

  private readonly logger: Logger;

  constructor(options: SagaOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Saga"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._causationList = options.causationList || [];
    this._destroyed = options.destroyed || false;
    this._messagesToDispatch = options.messagesToDispatch || [];
    this._revision = options.revision || 0;
    this._state = options.state || {};
  }

  public get causationList(): Array<string> {
    return this._causationList;
  }
  public set causationList(_) {
    throw new IllegalEntityChangeError();
  }

  public get messagesToDispatch(): Array<Command> {
    return this._messagesToDispatch;
  }
  public set messagesToDispatch(_) {
    throw new IllegalEntityChangeError();
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get revision(): number {
    return this._revision;
  }
  public set revision(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public destroy(): void {
    this.logger.debug("destroying");

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._destroyed = true;
  }

  public dispatch(causation: DomainEvent, command: Command): void {
    this.logger.debug("dispatching command", { causation, command });

    validateSchemaSync(SAGA_DISPATCH_SCHEMA, { command, causation });

    if (!(command instanceof Command)) {
      throw new MessageTypeError(command, Command);
    }

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Command(
        merge(
          {
            aggregate: {
              context: this.context,
            },
          },
          command,
        ),
        causation,
      ),
    );
  }

  public mergeState(data: Record<string, any>): void {
    this.logger.debug("merging state", { data });

    validateSchemaSync(SAGA_UPDATE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    merge(this._state, data);
  }

  public setState(path: string, value: AnyType): void {
    this.logger.debug("setting state", { path, value });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    set(this._state, path, value);
  }

  public timeout(
    causation: DomainEvent,
    name: string,
    data: Record<string, any>,
    delay: number,
  ): void {
    this.logger.debug("dispatching timeout", { causation, name, data, delay });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Timeout(
        {
          name,
          data,
          aggregate: {
            id: this.id,
            name: this.name,
            context: this.context,
          },
          delay,
        },
        causation,
      ),
    );
  }

  public toJSON(): SagaData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      causationList: cloneDeep(this.causationList),
      messagesToDispatch: cloneDeep(this.messagesToDispatch),
      destroyed: this.destroyed,
      revision: this.revision,
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: SagaIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
