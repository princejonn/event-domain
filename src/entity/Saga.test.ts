import MockDate from "mockdate";
import { Saga } from "./Saga";
import { IllegalEntityChangeError, MessageTypeError, SagaDestroyedError } from "../error";
import {
  commandWithData,
  commandWithDestroy,
  eventWithData,
  eventWithDestroy,
  logger,
  sagaIdentifier,
} from "../test";

MockDate.set("2020-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("Saga", () => {
  describe("when creating a saga with camelCase name", () => {
    it("should create saga", () => {
      expect(
        new Saga(
          {
            id: "id",
            name: "camelCasedName",
            context: "default",
          },
          logger,
        ),
      ).toMatchSnapshot();
    });
  });

  describe("when creating a saga with non-camelCase name", () => {
    it("should create saga", () => {
      expect(
        () =>
          new Saga(
            {
              id: "id",
              name: "snake_case_name",
              context: "default",
            },
            logger,
          ),
      ).toThrow(new Error("snake_case_name is not in camelCase"));
    });
  });

  describe("when setting state", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          state: { OBJ: { VAL: "overwrite" } },
        },
        logger,
      );

      saga.setState("OBJ.OBJ.VAL", 12345);
      saga.setState("OBJ.VAL", "value");
      saga.setState("OBJ.OBJ.OBJ", { objectData: true });
    });

    it("should set state and overwrite old values", () => {
      expect(saga.state.OBJ.VAL).toBe("value");
    });

    it("should set state with new values", () => {
      expect(saga.state.OBJ.OBJ.VAL).toBe(12345);
      expect(saga.state.OBJ.OBJ.OBJ.objectData).toBe(true);
    });
  });

  describe("when merging state", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          state: { OBJ: { VAL: "exists", OVER: "write" } },
        },
        logger,
      );

      saga.mergeState({ OBJ: { ADD: "value", OVER: "value", OBJ: { VAL: 12345 } } });
    });

    it("should keep unchanged values", () => {
      expect(saga.state.OBJ.VAL).toBe("exists");
    });

    it("should overwrite new values", () => {
      expect(saga.state.OBJ.OVER).toBe("value");
    });

    it("should add new values", () => {
      expect(saga.state.OBJ.ADD).toBe("value");
      expect(saga.state.OBJ.OBJ.VAL).toBe(12345);
    });
  });

  describe("when destroying a saga", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);

      saga.destroy();
    });

    it("should set saga as destroyed", () => {
      expect(saga.destroyed).toBe(true);
    });
  });

  describe("when dispatching a command", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);

      saga.dispatch(eventWithData, commandWithDestroy);
    });

    it("should add the command in the dispatch list", () => {
      expect(saga.messagesToDispatch).toMatchSnapshot();
    });
  });

  describe("when timing out an event", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);

      saga.timeout(eventWithData, eventWithDestroy.name, eventWithDestroy.data, 999);
    });

    it("should add the command in the dispatch list", () => {
      expect(saga.messagesToDispatch).toMatchSnapshot();
    });
  });

  describe("when trying to dispatch an event", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw MessageTypeError", () => {
      expect(() => saga.dispatch(eventWithData, eventWithDestroy)).toThrow(
        MessageTypeError,
      );
    });
  });

  describe("when trying to merge state on a destroyed saga", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          destroyed: true,
        },
        logger,
      );
    });

    it("should throw SagaDestroyedError", () => {
      expect(() => saga.mergeState({ data: true })).toThrow(SagaDestroyedError);
    });
  });

  describe("when trying to set state on a destroyed saga", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          destroyed: true,
        },
        logger,
      );
    });

    it("should throw SagaDestroyedError", () => {
      expect(() => saga.setState("path", "value")).toThrow(SagaDestroyedError);
    });
  });

  describe("when trying to destroy a destroyed saga", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          destroyed: true,
        },
        logger,
      );
    });

    it("should throw SagaDestroyedError", () => {
      expect(() => saga.destroy()).toThrow(SagaDestroyedError);
    });
  });

  describe("when trying to dispatch a command on a destroyed saga", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(
        {
          ...sagaIdentifier,
          destroyed: true,
        },
        logger,
      );
    });

    it("should throw SagaDestroyedError", () => {
      expect(() => saga.dispatch(eventWithData, commandWithData)).toThrow(
        SagaDestroyedError,
      );
    });
  });

  describe("when trying to set causationList", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        saga.causationList = [];
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set destroyed", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        saga.destroyed = true;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set messagesToDispatch", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        saga.messagesToDispatch = [];
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set revision", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        saga.revision = 2;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set state", () => {
    let saga: Saga<any>;

    beforeAll(() => {
      saga = new Saga(sagaIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        saga.state = {};
      }).toThrow(IllegalEntityChangeError);
    });
  });
});
