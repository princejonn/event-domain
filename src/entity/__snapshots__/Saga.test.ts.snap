// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga when creating a saga with camelCase name should create saga 1`] = `
Object {
  "causationList": Array [],
  "context": "default",
  "destroyed": false,
  "id": "id",
  "messagesToDispatch": Array [],
  "name": "camelCasedName",
  "revision": 0,
  "state": Object {},
}
`;

exports[`Saga when dispatching a command should add the command in the dispatch list 1`] = `
Array [
  Command {
    "aggregate": Object {
      "context": "defaultContext",
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "name": "aggregateName",
    },
    "causationId": "fc2bd748-91e3-46ce-b5e5-6cdd2184122f",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {},
    "delay": 0,
    "id": "fc2bd748-91e3-46ce-b5e5-6cdd2184122f",
    "mandatory": true,
    "name": "commandWithDestroy",
    "timestamp": 2020-01-01T12:00:00.000Z,
    "type": "command",
  },
]
`;

exports[`Saga when timing out an event should add the command in the dispatch list 1`] = `
Array [
  Timeout {
    "aggregate": Object {
      "context": "defaultContext",
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "name": "sagaName",
    },
    "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {},
    "delay": 999,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "mandatory": true,
    "name": "eventWithDestroy",
    "timestamp": 2020-01-01T08:00:00.000Z,
    "type": "timeout",
  },
]
`;
